lib/utils/eol-conversion.ts

```diff
@@ -0,0 +1,110 @@
+/**
+ * @license
+ * Copyright 2019-2020 Balena Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import mmmagic = require('mmmagic');
+import fs = require('mz/fs');
+import Logger = require('./logger');
+
+const globalLogger = Logger.getLogger();
+
+// Define file size threshold (bytes) over which analysis/conversion is not performed.
+const LARGE_FILE_THRESHOLD = 100 * 1000 * 1000;
+
+/**
+ * Attempt to detect the file encoding of a data buffer.
+ * @param data Raw buffer to analyse
+ */
+async function detectEncoding(data: Buffer): Promise<string> {
+	// Instantiate mmmagic for mime encoding analysis
+	const magic = new mmmagic.Magic(mmmagic.MAGIC_MIME_ENCODING);
+
+	// Promisify magic.detect
+	// For some reason, got 'Illegal Invocation' when using:
+	//   const detectEncoding = promisify(magic.detect);
+	return new Promise((resolve, reject) => {
+		magic.detect(data, (err, res) => {
+			if (err) {
+				return reject(err);
+			}
+			return resolve(res);
+		});
+	});
+}
+
+/**
+ * Drop-in replacement for promisified fs.readFile(<string>)
+ * Attempts to convert EOLs from CRLF to LF for supported encodings,
+ * or otherwise logs warnings.
+ * @param filepath
+ * @param convertEol When true, performs conversions, otherwise just warns.
+ */
+export async function readFileWithEolConversion(
+	filepath: string,
+	convertEol: boolean,
+): Promise<Buffer> {
+	// The list of encodings to convert is intentionally conservative for now
+	const convertableEncodings = ['ascii', 'utf-8'];
+
+	const fileBuffer = await fs.readFile(filepath);
+
+	// Skip processing of very large files
+	const fileStats = await fs.stat(filepath);
+	if (fileStats.size > LARGE_FILE_THRESHOLD) {
+		globalLogger.logWarn(`CRLF detection skipped for large file: ${filepath}`);
+		return fileBuffer;
+	}
+
+	// Analyse encoding
+	let encoding = await detectEncoding(fileBuffer);
```

Comment:
> additional copies that co-exist in memory

Btw, I think that [copy-on-write](https://en.wikipedia.org/wiki/Copy-on-write) doesn't apply because of the encoding conversion (UTF-8), for example:
```
let euro = '€';
let a = Buffer.from(euro, 'utf-8');
let b = Buffer.from(euro, 'ascii');
euro.length → result is 1
a.length → result is 3
b.length → result is 1
```

And the OS kernel behaviour might be different in Linux, macOS and Windows.

Reply:
Many good points here. This whole thing is one of the reasons I added the skip for large files.

> So I suggest that we give the filename rather than a Buffer to detectEncoding():
Hard to say which more efficient without tests, but you're probably right.

> Then, only after mmmagic decided that the file is a text file, we'd load the (whole) file to a Buffer in this function
The function must do a readFile to buffer in all circumstances (because is it acting as a drop-in replacement for readFile in the consuming code) so whether we do this earlier or later makes no difference IMO. (and in fact the initial size test must return the buffer if file is large).

> Instead of fileDataString.includes('\r\n'), we could run fileBuffer.includes('\r\n') or fileBuffer.indexOf('\r\n')
Yeah right.

> Just for fun, I've written the function below, convertEolInPlace()
Very elegant algo!  I'd considered modifying the buffer directly, but got dissuaded by the differing byte-lengths of CR and CRLF.  This is much better, thanks!

> As a bonus, the custom function converts single CR characters (not part of a CRLF sequence) to single LF characters.
Ha, I think I'd consider this a side effect, rather than a bonus ;)  I imagine we would almost never see a line CR in the wild, but if it's there I don't think we should assume it can be safely converted to LF.  Who knows what they might be using it for, I can't think of anything really useful, but maybe some text display hack, maybe they have a rPi connect to an 8-bit computer ;)   I'll take a look at modifying the algo without massive changes.

Reply:
> The function must do a readFile to buffer in all circumstances (because is it acting as a drop-in replacement for readFile in the consuming code) so whether we do this earlier or later makes no difference IMO. 

Right! My suggestion of using `magic.detectFile(filepath, ...)` instead of `magic.detect(data, ...)` doesn't make sense. :-)